### PR TITLE
Improve how IndexSettingProvider instances are managed

### PR DESCRIPTION
**Pull request by martijnvg**
by adding a IndexSettingProviders class that keeps track of the various implementations
and can be passed around to other node level services.

Also moved IndexSettingProvider interface to org.elasticseach.index package
(because it is an index level component) and added more java docs.

Follow-up from #81532

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexSettingProviders.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index;
+
+import java.util.Collections;
+import java.util.Set;
+
+public final class IndexSettingProviders {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/IndexSettingProviders.java:14]: Can you add a javadoc description for this class?

**Review by dakrone — APPROVED** [review]
LGTM, I think this is nicer since it makes everything immutable, thanks for doing this!

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Pull request merged**